substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java

```diff
@@ -0,0 +1,548 @@
+     * {@link InvocationResultInline}.
+     */
+    static final class CallSite {
+        final AnalysisMethod caller;
+        final int bci;
+
+        CallSite(AnalysisMethod caller, int bci) {
+            this.caller = caller;
+            this.bci = bci;
+        }
+
+        @Override
+        public int hashCode() {
+            return caller.hashCode() * 31 + bci;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (this == obj) {
+                return true;
+            } else if (obj == null || getClass() != obj.getClass()) {
+                return false;
+            }
+            CallSite other = (CallSite) obj;
+            return bci == other.bci && caller.equals(other.caller);
+        }
+
+        @Override
+        public String toString() {
+            return caller.format("%h.%n(%p)") + "@" + bci;
+        }
+    }
+
+    static class InvocationResult {
+        static final InvocationResult ANALYSIS_TOO_COMPLICATED = new InvocationResult();
+
+    }
+
+    public static class InvocationResultInline extends InvocationResult {
+        final CallSite site;
+        final AnalysisMethod callee;
+        final Map<CallSite, InvocationResultInline> children;
+
+        public InvocationResultInline(CallSite site, AnalysisMethod callee) {
+            this.site = site;
+            this.callee = callee;
+            this.children = new HashMap<>();
+        }
+
+        @Override
+        public String toString() {
+            return append(new StringBuilder(), "").toString();
+        }
+
+        private StringBuilder append(StringBuilder sb, String indentation) {
+            sb.append(site).append(" -> ").append(callee.format("%h.%n(%p)"));
+            String newIndentation = indentation + "  ";
+            for (InvocationResultInline child : children.values()) {
+                sb.append(System.lineSeparator()).append(newIndentation);
+                child.append(sb, newIndentation);
+            }
+            return sb;
+        }
+    }
+}
+
+/**
+ * This class detects what can be inlined.
+ */
+class TrivialMethodDetector {
+
+    final BigBang bb;
+    final HostedProviders providers;
+    final GraphBuilderConfiguration prototypeGraphBuilderConfig;
+    final OptionValues options;
+    final DebugContext debug;
+
+    TrivialMethodDetector(BigBang bb, HostedProviders providers, GraphBuilderConfiguration originalGraphBuilderConfig, OptionValues options, DebugContext debug) {
+        this.bb = bb;
+        this.debug = debug;
+        this.prototypeGraphBuilderConfig = makePrototypeGraphBuilderConfig(originalGraphBuilderConfig);
+        this.options = options;
+        this.providers = providers;
+    }
+
+    private static GraphBuilderConfiguration makePrototypeGraphBuilderConfig(GraphBuilderConfiguration originalGraphBuilderConfig) {
+        GraphBuilderConfiguration result = originalGraphBuilderConfig.copy();
+
+        result.getPlugins().clearParameterPlugin();
```

Comment:
Is this overly destructive? Are there any other parameter plugins that we might need?

Reply:
I tested the latest version, remove or leave them, either one is fine because two of them are in the compiler package, and one is for intrinsification (we are not interfering with any intrinsic).